### PR TITLE
[3.10] SEARCH-461 cache only leaders

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose

Adds option "--arangosearch.columns-cache-only-leader". Used only on DBServers EE.
Default is false. If set to true only leader shards have ArangoSearch caches enabled - this will reduce RAM usage.
In case of failover happens - in background caches are populated for the new leader. But some queries that a run at same time as failover may still run without caches.

Enterprise PR: https://github.com/arangodb/enterprise/pull/1251

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/IResearch/IResearchDataStore.cpp [hunk 1]
```diff
@@ -974,6 +989,35 @@ Result IResearchDataStore::commitUnsafeImpl(
       LOG_TOPIC("4cb66", DEBUG, TOPIC) << "Commit started";
     }
 #endif
+
+
+    auto forceOpen = [&]() -> bool {
+      bool force{false};
+#ifdef USE_ENTERPRISE
+      if (ServerState::instance()->isDBServer() &&
```
Comment (MBkkt) [inline arangod/IResearch/IResearchDataStore.cpp:997]: I think we always could use columnsCacheOnlyLeaders/_useSearchCache

File: arangod/Cluster/ClusterInfo.cpp [hunk 2]
```diff
@@ -5726,6 +5726,30 @@ std::shared_ptr<std::vector<ServerID> const> ClusterInfo::getResponsibleServer(
...
+  if (it != _shardIds.end()) {
+    auto serverList = (*it).second;
+    if (!serverList || serverList->empty()) {
+      return ShardLeadership::kUnclear;
+    }
+    if (!(*serverList)[0].empty() && (*serverList)[0][0] == '_') {
+      // This is a temporary situation in which the leader has already
+      return ShardLeadership::kUnclear;
+    } else {
+      return serverList->at(0) == server ? ShardLeadership::kLeader
```
Comment (MBkkt) [inline arangod/Cluster/ClusterInfo.cpp:5746]:
```suggestion
      return (*serverList)[0] == server ? ShardLeadership::kLeader
```

**Review by MBkkt — APPROVED** [review]

File: arangod/Cluster/ClusterInfo.cpp [hunk 3]
```diff
@@ -5726,6 +5726,30 @@ std::shared_ptr<std::vector<ServerID> const> ClusterInfo::getResponsibleServer(
...
+  }
+  READ_LOCKER(readLocker, _currentProt.lock);
+  auto it = _shardIds.find(shard);
+
+  if (it != _shardIds.end()) {
+    auto const& serverList = (*it).second;
+    if (!serverList || serverList->empty()) {
+      return ShardLeadership::kUnclear;
+    }
+    if (!(*serverList)[0].empty() && (*serverList)[0][0] == '_') {
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp:5742]: Slightly simpler:
```suggestion
    if ((*serverList)[0].starts_with('_')) {
```

File: arangod/Cluster/ClusterInfo.cpp [hunk 4]
```diff
@@ -5726,6 +5726,30 @@ std::shared_ptr<std::vector<ServerID> const> ClusterInfo::getResponsibleServer(
...
+      // This is a temporary situation in which the leader has already
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp:5743]: the comment seems to be cut off. The comment appears in other places in this file, and there the comment ends with `in which the leader has already resigned.`.

File: arangod/Cluster/ClusterInfo.cpp [hunk 5]
```diff
@@ -5726,6 +5726,31 @@ std::shared_ptr<std::vector<ServerID> const> ClusterInfo::getResponsibleServer(
   return std::make_shared<std::vector<ServerID>>();
 }
 
+ClusterInfo::ShardLeadership ClusterInfo::getShardLeadership(
+    ServerID server, ShardID shard) const {
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp:5730]: note: this line also needs adjustment after the corresponding header was changed.

File: arangod/IResearch/IResearchDataStore.cpp [hunk 6]
```diff
@@ -1037,9 +1078,15 @@ Result IResearchDataStore::commitUnsafeImpl(
...
     }
 #endif
-    // get new reader
-    auto reader = _dataStore._reader.reopen(_format);
 
+#ifdef USE_ENTERPRISE
+    auto reader = forceOpen() ? irs::directory_reader::open(
+                                    *(_dataStore._directory), _format,
+                                    _dataStore._readerOptions)
+                              : _dataStore._reader.reopen(_format);
```
Comment (jsteemann) [inline arangod/IResearch/IResearchDataStore.cpp:1086]: same here: how does the reader know about the value of `_useSearchCache`?

File: arangod/IResearch/IResearchDataStore.cpp [hunk 7]
```diff
@@ -986,6 +1020,13 @@ Result IResearchDataStore::commitUnsafeImpl(
       _lastCommittedTickOne = lastTickBeforeCommitOne;
       _lastCommittedTickTwo = lastTickBeforeCommitOne;
       impl.tick(_lastCommittedTickOne);
+#ifdef USE_ENTERPRISE
+      // get new reader
+      if (forceOpen()) {
+        _dataStore._reader = irs::directory_reader::open(
+            *(_dataStore._directory), _format, _dataStore._readerOptions);
+      }
```
[the remark this answers is not in the document]
Comment (Dronplane) [reply arangod/IResearch/IResearchDataStore.cpp:1028]: Reference to `_useSearchCache` is stored in reader options. Implicitly - as lambda capture.

File: arangod/IResearch/IResearchDataStore.cpp (hunk 6 above)
Comment (Dronplane) [reply arangod/IResearch/IResearchDataStore.cpp:1086]: Reference to _useSearchCache is stored in reader options. Implicitly - as lambda capture.

**Review by jsteemann — APPROVED** [review]
LGTM

**Review by KVS85 — APPROVED** [review]

**Pull request merged**